File: arrays_hashing/top_k_frequent_elements/js/top_k_frequent_elements.js
const nums = [ 1, 1, 1, 2, 2, 3 ];
const k = 2;

const topKFrequent = (nums, k) => {
  let map = new Map(); 
  let top = new Map();
  let result = new Array();

  // record frequency of number
  for(let n of nums) {
    let value = map.get(n) || [];
    value++;
    map.set(n, value);
  }

  // index them by frequency
  for(let m of map) {
    let value = top.get(m[1]) || [];
    value.push(m[0]);
    top.set(m[1], value);
  }

  // sort the hash keys O(Nlog(n))
  let ranked_keys = Array.from(top)
    .map(val => val[0])
    .sort((a, b) => b - a);

  // push top k items to result list
  for(let i = 0; i < top.size; i++) {
    let elem = top.get(ranked_keys[i]);

    for(let subElem of elem) {
      if(k == 0) return result; k--;

      result.push(subElem);
    }
  }

  return result;
};

console.log(topKFrequent(nums, k));

File: arrays_hashing/top_k_frequent_elements/cpp/top_k_frequent_elements.cpp
#include<iostream>
#include<vector>
#include<unordered_map>
#include<map>

using std::cout;
using std::vector;

// Time O(n)
// Space O(n)

class Solution {
public:
    static vector<int> topKFrequent(const vector<int>& nums, int k) {
        std::unordered_map<int, int> map; // c++ maps are zero initialized
        std::map<int, vector<int>> top;   // ordered map of highest frequency nums
        vector<int> result = {};
    
        // rank num frequency O(n)
        for(const auto& n : nums) {
            map[n]++;
        }

        // order num frequency into vectors O(n)
        for(const auto& n : map) {
            top[n.second].push_back(n.first);
        }

        // reverse iterate to get highest frequency in O(k)
        for(auto rit = top.rbegin(); rit != top.rend(); ++rit) {
            for(auto& n : rit->second) {
                if(!k) break; k--;
                result.push_back(n);
            }
        }

        return result;
    }
};

int main() {
    vector<int> nums = { 1, 1, 1, 2, 2, 3 };
    int k = 1;

    auto result = Solution::topKFrequent(nums, k);

    for(auto& n : result) {
        cout << n << '\n';
    }
}
